### PR TITLE
Group modules by package in the sidebar

### PR DESCRIPTION
**Pull request by klntsky**
2020-07-06-212400_432x545_escrotum
2020-07-06-212411_296x488_escrotum

### CODE REVIEW

**Comment by klntsky** [comment]
Known issues:

- if there are no definitions in a module, it will not show up in the listing (Right now we are getting package info by looking at the code locations of the definitions; so no definitions is indistinguishable from a non-existing module)
- visual design

To consider:

- Do we want "expand all" functionality in the package tree?

File: src/Docs/Search/App/Sidebar.purs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                          , mode: GroupByPackage
+                          , moduleNames
+                          }
+    , render
+    , eval: H.mkEval $ H.defaultEval { handleAction = handleAction }
+    }
+    where
+      moduleNames = Array.sort $ Array.fromFoldable $ foldr Set.union mempty moduleIndex
+
+      -- Convert `ModuleIndex` to the appropriate format.
```
Comment (f-f) [inline src/Docs/Search/App/Sidebar.purs]: Could you expand the comment on the "appropriate format"? Why is it better than the input format?

File: src/Docs/Search/App/Sidebar.purs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+             , HE.onChecked $ Just <<< ToggleGrouping
+             ]
+
+  , HH.text " "
+  , HH.label [ HP.for "group-modules__input"
+             , HP.id_ "group-modules__label"
+             ]
+    [ HH.text " GROUP BY PACKAGE" ]
+
+  , if mode == GroupByPackage
```
Comment (m-bock) [inline src/Docs/Search/App/Sidebar.purs]: Could we maybe extract the `HH.ul_` as it appears in both branches of the `if .. then .. else`?

File: src/Docs/Search/App/Sidebar.purs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+      HH.li [ HE.onClick $ const $ Just $ ToggleCollapse packageName
+            , HP.classes [ wrap $ if status == Expanded
+                                  then "li-expanded-package"
+                                  else "li-collapsed-package"
+                         , wrap "li-package" ]
+            ]
+
+      if status == Expanded
+      then [ HH.text packageName
```
Comment (m-bock) [inline src/Docs/Search/App/Sidebar.purs]: Maybe the same here, `HH.text packageName` is not part of the conditions...

File: src/Docs/Search/App/Sidebar.purs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    packageList :: Array (String /\ Set ModuleName /\ PackageEntryState)
+    packageList =
+      first (Array.foldMap String.singleton) <$> (
+        Trie.toUnfoldable expansions :: Array (Array Char /\ Set ModuleName /\ PackageEntryState)
+      )
+
+
+-- Some optics:
+
+_expansions :: forall a b rest.  (a -> b) -> { expansions :: a | rest } -> { expansions :: b | rest }
```
Comment (m-bock) [inline src/Docs/Search/App/Sidebar.purs]: extra space, or is that on purpose?.. don't know..

File: src/Docs/Search/Types.purs [hunk 5]
```diff
@@ -0,0 +1,5 @@
+module Docs.Search.Types where
+
+type ModuleName = String
```
Comment (m-bock) [inline src/Docs/Search/Types.purs:3]: Maybe we could consider using newtypes here..

**Review by m-bock — COMMENTED** [review]
very cool work, a feature I've been really looking forward to. thx!!

File: src/Docs/Search/Types.purs (hunk 5 above)
Comment (klntsky) [reply src/Docs/Search/Types.purs:3]: I'll do it in a separate PR. The best set of types here would be:

```purescript
newtype ModuleName = ModuleName String
newtype PackageName = PackageName String
data PackageInfo = Package PackageName | Builtin | LocalPackage | UnknownPackage
```

**Review by f-f — APPROVED** [review]
I left a small request for more docs, but this looks amazing otherwise, great great work! :clap:

File: src/Docs/Search/App/Sidebar.purs (hunk 1 above)
Comment (klntsky) [reply src/Docs/Search/App/Sidebar.purs]: Good point. Initially I thought that a trie would be better, cause Map degrades when used with many similar strings (package names tend to have common prefixes). But now I don't even do lookups, (thanks to @paf31's suggestion to use `<summary>`), so a trie is not needed.

**Comment by klntsky** [comment]
Oh and also, we should preserve the state of the checkbox between page loads.

File: src/Docs/Search/App/Sidebar.purs [hunk 6]
```diff
@@ -10,21 +11,26 @@ import Data.Lens.Record (prop)
...
 import Halogen.HTML as HH
 import Halogen.HTML.Events as HE
 import Halogen.HTML.Properties as HP
+import Web.HTML as HTML
+import Web.HTML.Window as Window
+import Web.Storage.Storage as Storage
 
 
-data Action = ToggleGrouping Boolean
+data Action = ToggleGrouping Mode
```
Comment (f-f) [inline src/Docs/Search/App/Sidebar.purs]:
```suggestion
data Action = ToggleGrouping GroupingMode
```

**Review by f-f — APPROVED** [review]
:100: :clap:

**Pull request merged**